core/src/test/scala/unit/kafka/log/LogCleanerTest.scala

```diff
@@ -699,28 +699,70 @@ class LogCleanerTest extends JUnitSuite {
+    groups = cleaner.groupSegmentsBySize(log.logSegments, maxSize = Int.MaxValue, maxIndexSize = Int.MaxValue, log.logSegments.last.index.lastOffset + 1)
     assertEquals(2, groups.size)
     checkSegmentOrder(groups)
 
     // append more messages, creating new segments, further grouping should still occur
     while (log.numberOfSegments < 4)
       log.appendAsLeader(TestUtils.singletonRecords(value = "hello".getBytes, key = "hello".getBytes), leaderEpoch = 0)
 
-    groups = cleaner.groupSegmentsBySize(log.logSegments, maxSize = Int.MaxValue, maxIndexSize = Int.MaxValue)
+    groups = cleaner.groupSegmentsBySize(log.logSegments, maxSize = Int.MaxValue, maxIndexSize = Int.MaxValue, log.logSegments.last.index.lastOffset + 1)
     assertEquals(log.numberOfSegments - 1, groups.size)
     for (group <- groups)
       assertTrue("Relative offset greater than Int.MaxValue", group.last.index.lastOffset - group.head.index.baseOffset <= Int.MaxValue)
     checkSegmentOrder(groups)
   }
 
+  /** 
+   * Following the loading of a log segment where the index file is zero sized,
+   * the index returned would be the base offset.  Sometimes the log file would
+   * contain data with offsets in excess of the baseOffset which would cause
+   * the log cleaner to group together segments with a range of > Int.MaxValue
+   * this test replicates that scenario to ensure that the segments are grouped
+   * correctly.
+   */
+  @Test
+  def testSegmentGroupingFollowingLoadOfZeroIndex(): Unit = {
+    val cleaner = makeCleaner(Int.MaxValue)
+
+    val logProps = new Properties()
+    logProps.put(LogConfig.SegmentBytesProp, 400: java.lang.Integer)
+    logProps.put(LogConfig.IndexIntervalBytesProp, 1: java.lang.Integer)
+
+    val log = makeLog(config = LogConfig.fromProps(logConfig.originals, logProps))
+
+    // fill up first segment
+    while (log.numberOfSegments == 1)
+      log.appendAsLeader(TestUtils.singletonRecords(value = "hello".getBytes, key = "hello".getBytes), leaderEpoch = 0)
+
+    // forward offset and append message to next segment at offset Int.MaxValue
+    val records = messageWithOffset("hello".getBytes, "hello".getBytes, Int.MaxValue -1)
+    log.appendAsFollower(records)
+    log.appendAsLeader(TestUtils.singletonRecords(value = "hello".getBytes, key = "hello".getBytes), leaderEpoch = 0)
+    assertEquals(Int.MaxValue, log.activeSegment.index.lastOffset)
+   
+    //mimic the effect of loading an index with no indexes - last offset gets set to baseOffset
+    log.activeSegment.index.truncateTo(log.activeSegment.baseOffset)
+
+    assertTrue("lastOffset should be less than the true last offset", log.activeSegment.index.lastOffset < Int.MaxValue)
+   
+    // grouping should result in a two groups because the second segment contains an index > MaxInt
+    val groups = cleaner.groupSegmentsBySize(log.logSegments, maxSize = Int.MaxValue, maxIndexSize = Int.MaxValue, log.logSegments.last.baseOffset + Int.MaxValue)
```

Comment:
We can use log.logEndOffset instead of log.logSegments.last.index.lastOffset + 1.